src/lib.rs

```diff
@@ -1,7 +1,552 @@
+    res_G_0s: Vec<secp256k1::Scalar>,
+    /// Responses for proofs that a bit of the ed25519 scalar is equal to 0.
+    res_H_0s: Vec<ed25519::Scalar>,
+    /// Responses for proofs that a bit of the secp256k1 scalar is equal to 1.
+    res_G_1s: Vec<secp256k1::Scalar>,
+    /// Responses for proofs that a bit of the ed25519 scalar is equal to 1.
+    res_H_1s: Vec<ed25519::Scalar>,
+    /// Blinder for the "overall" Pedersen Commitment of the secp256k1 scalar.
+    ///
+    /// Calculation: sum of `r_i * 2^i`, where `i` is the index of the bit and
+    /// `r_i` its blinder.
+    r: secp256k1::Scalar,
+    /// Blinder for the "overall" Pedersen Commitment of the ed25519 scalar.
+    ///
+    /// Calculation: sum of `s_i * 2^i`, where `i` is the index of the bit and
+    /// `s_i` its blinder.
+    s: ed25519::Scalar,
+}
+
+impl Proof {
+    /// Prove knowledge of `witness` across secp256k1 and ed25519. In other
+    /// words, prove knowledge of the value of `witness`, whilst also proving
+    /// that it has an equivalent representation on secp256k1 and ed25519.
+    ///
+    /// # High-level strategy
+    ///
+    /// 1. Decompose `witness` into bits.
+    /// 2. For each bit, generate a Pedersen Commitment of the form
+    /// `C_G_i = bit_i * G + r_i * G_PRIME`, where `r_i` is a blinder, and `G`
+    /// and `G_PRIME` are generators of secp256k1.
+    /// 3. For each bit, also generate a Pedersen Commitment of the form
+    /// `C_H_i = bit * H + s_i * H_PRIME`, where `s_i` is a blinder, and `H` and
+    /// `H_PRIME` are generators of ed25519.
+    /// 4. Leveraging the [_composition of Sigma-protocols_], for every bit
+    /// prove knowledge of
+    ///
+    /// - the discrete logarithm w.r.t. `G_PRIME` of `C_G_i` AND the discrete
+    ///   logarithm w.r.t `H_PRIME` of `C_H_i`; OR
+    /// - the discrete logarithm w.r.t. `G_PRIME` of `C_G_i - G` AND the
+    ///   discrete logarithm w.r.t `H_PRIME` of `C_H_i - H`.
+    ///
+    /// This proves that each bit is the same, that each Pedersen Commitment is
+    /// to either 0 or 1, and knowledge of the corresponding blinders (`r_i` and
+    /// `s_i`) for each bit.
+    ///
+    /// 5. Prove that the sum of all `C_G_i * 2^i` minus `r * G_PRIME` is equal
+    /// to the public value `witness * G`, and that the sum of all `C_H_i * 2^i`
+    /// minus `s * H_PRIME` is equal to the public value `witness * H`.
+    ///
+    /// To this end, the proof contains `r` equal to the sum `r_i * 2^i` for all
+    /// `i` and the `s` equal to the sum `s_i * 2^i` for all `i`. The verifier
+    /// will simply operate as indicated above and verify that the equalities
+    /// hold.
+    ///
+    /// [_composition of Sigma-protocols_]: https://www.win.tue.nl/~berry/CryptographicProtocols/LectureNotes.pdf
+    pub fn new<R: RngCore + CryptoRng>(rng: &mut R, witness: &Scalar) -> Proof {
+        let bits = witness.bits();
+
+        let mut C_G_is = Vec::new();
```

Comment:
It should be possible to implement this without using `Vec` and alloc. See `generic_array`.

Reply:
Very cool tip! Didn't know about `generic_array`. Definitely going to do this.